### ISSUE TITLE
Crash with warm cash and inheriting base classes in combination with a custom base

### ISSUE THREAD

**Issue by aberres** [sqlalchemy/sqlalchemy#7034]
**Describe the bug**

The reproduction of this bug is somewhat tricky. I'll try to provide all the information I have.

Please see the stripped files below. We need two files with inheriting base classes and  a custom base passed to `declarative_base`.

Here is what I do to get the crash

- Run mypy with cleared cache - everything is fie 
- Change the file containing `Base2` - observe crash
- Delete the cache and rerun mypy - things are fine

So I guess when `Base` is already in the cache on a run where `Base2` is rechecked, something is missing.

**Expected behavior**

No crash.

**To Reproduce**

*modelbase.py*
```
from sqlalchemy.ext.declarative import declarative_base,

class _CustomBase:
    pass

# No crash without the custom base class
Base = declarative_base(cls=_CustomBase)
```
*another_file.py*
```python
from mypackage.modelbase import Base

class Base2(Base):
    __abstract__ = True
```

**Error**

```
version: 0.910
Traceback (most recent call last):
  File "mypy/semanal.py", line 4872, in accept
  File "mypy/nodes.py", line 950, in accept
  File "mypy/semanal.py", line 1056, in visit_class_def
  File "mypy/semanal.py", line 1134, in analyze_class
  File "mypy/semanal.py", line 1143, in analyze_class_body_common
  File "mypy/semanal.py", line 1203, in apply_class_plugin_hooks
  File "/Users/armin/venv/weplan3.9/lib/python3.9/site-packages/sqlalchemy/ext/mypy/plugin.py", line 199, in _base_cls_hook
    decl_class._scan_declarative_assignments_and_apply_types(ctx.cls, ctx.api)
  File "/Users/armin/venv/weplan3.9/lib/python3.9/site-packages/sqlalchemy/ext/mypy/decl_class.py", line 67, in _scan_declarative_assignments_and_apply_types
    assert cls_metadata.is_mapped
```

**Versions.**
 - OS: MacOS
 - Python: 3.9
 - SQLAlchemy: 1.4.22
 - mypy: 0.910
 - SQLAlchemy2-stubs: 0.0.2a15

**Comment by zzzeek** [comment]
hey there -

please update to 1.4.23, the mypy plugin has been refactored and that line of code is no longer present.

**Comment by zzzeek** [comment]
confirmed this issue is fixed, i was able to reproduce in 1.4.22, the refactoring for 1.4.23 resolves.

**Comment by aberres** [comment]
Oooh, sorry for the noise and thanks.

Looks good!
